agentMET4FOF/metrological_agents.py

```diff
@@ -356,4 +353,833 @@ def agent_loop(self):
+                t_data_arr2d[:, i_sensor] = data_arr[:, 0]
+                ut_data_arr2d[:, i_sensor] = data_arr[:, 1]
+                x_data_arr2d[:, i_sensor] = data_arr[:, 2]
+                ux_data_arr2d[:, i_sensor] = data_arr[:, 3]
+                i_sensor = i_sensor + 1
+
+            if self.calc_type == "lcs":
+                data = np.full(shape=(self.n_pr, 4), fill_value=np.nan)
+                for i_pnt in range(self.n_pr):
+                    y_arr = np.array(x_data_arr2d[i_pnt, :])
+                    y_arr = y_arr.reshape((n_sensors, 1))
+                    vy_arr2d = np.zeros(shape=(n_sensors, n_sensors))
+                    for i_sensor in range(n_sensors):
+                        vy_arr2d[i_sensor, i_sensor] = np.square(ux_data_arr2d[i_pnt, i_sensor])
+
+                    n_sols, ybest, uybest, chi2obs, indkeep = self.calc_lcs(y_arr, vy_arr2d, self.problim)
+                    if n_sols == 1:  # time stamp is value of first sensor
+                        if isinstance(ybest, np.ndarray):
+                            ybest = ybest[0]
+                        data[i_pnt, :] = np.array([t_data_arr2d[i_pnt, 0], ut_data_arr2d[i_pnt, 0], ybest, uybest])
+                    else:  # only return the first solution
+                        data[i_pnt, :] = np.array([t_data_arr2d[i_pnt, 0], ut_data_arr2d[i_pnt, 0], ybest[0], uybest[0]])
+            elif self.calc_type == "lcss":
+                # lcss applied to one data vector (required input)
+                # Sum the signals to get one signal
+                x_data_arr = np.sum(x_data_arr2d, axis=1)
+                x_data_arr = x_data_arr.reshape((len(x_data_arr), 1))
+                ux2_data_arr = np.sum(np.square(ux_data_arr2d), axis=1)
+                vx_arr2d = np.zeros((self.n_pr, self.n_pr))
+                for i_val in range(self.n_pr):
+                    vx_arr2d[i_val, i_val] = ux2_data_arr[i_val]
+
+                n_sols, ybest, uybest, chi2obs, indkeep = \
+                    self.calc_lcss(self.a_arr, self.a_arr2d, x_data_arr, vx_arr2d, self.problim)
+                print('calc lcss finished')
+                print('n_sols: ', n_sols)
+                print('ybest: ', ybest)
+                print('uybest: ', uybest)
+                if n_sols == 1:  # time stamp is latest value
+                    if isinstance(ybest, np.ndarray):
+                        ybest = ybest[0]
+                    data = np.array([t_data_arr2d[-1, 0], ut_data_arr2d[-1, 0], ybest, uybest])
+                else:  # only return the first solution
+                    data = np.array([t_data_arr2d[-1, 0], ut_data_arr2d[-1, 0], ybest[0], uybest[0]])
+
+            # Send the data
+            if len(data.shape) == 1:
+                data = data.reshape((1, len(data)))
+
+            self.set_output_data(channel="default", data=data)
+            super().agent_loop()
+
+    def calc_consistent_estimates_no_corr(y_arr2d, uy_arr2d, prob_lim):
```

Comment:
This is a method, so the first parameter needs to be `self` or we have to make it a static function.